streams/src/test/java/org/apache/kafka/streams/integration/StreamsUpgradeTestIntegrationTest.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.kafka.streams.integration.utils.IntegrationTestUtils;
+import org.apache.kafka.streams.tests.StreamsUpgradeTest;
+import org.apache.kafka.test.IntegrationTest;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.time.Duration;
+import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.apache.kafka.common.utils.Utils.mkProperties;
+import static org.apache.kafka.streams.processor.internals.assignment.StreamsAssignmentProtocolVersions.LATEST_SUPPORTED_VERSION;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@Category(IntegrationTest.class)
+public class StreamsUpgradeTestIntegrationTest {
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(3);
+
+    @BeforeClass
+    public static void setup() {
+        IntegrationTestUtils.cleanStateBeforeTest(CLUSTER, 1, "data");
+    }
+
+    @Test
+    public void testVersionProbingUpgrade() throws InterruptedException {
+        final KafkaStreams kafkaStreams1 = StreamsUpgradeTest.buildStreams(mkProperties(
+            mkMap(
+                mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers())
+            )
+        ));
+        final KafkaStreams kafkaStreams2 = StreamsUpgradeTest.buildStreams(mkProperties(
+            mkMap(
+                mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers())
+            )
+        ));
+        final KafkaStreams kafkaStreams3 = StreamsUpgradeTest.buildStreams(mkProperties(
+            mkMap(
+                mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers())
+            )
+        ));
+        startSync(kafkaStreams1, kafkaStreams2, kafkaStreams3);
+
+        // first roll
+        kafkaStreams1.close();
+        final AtomicInteger usedVersion4 = new AtomicInteger();
+        final KafkaStreams kafkaStreams4 = buildFutureStreams(usedVersion4);
+        startSync(kafkaStreams4);
+        assertThat(usedVersion4.get(), is(LATEST_SUPPORTED_VERSION));
+
+        // second roll
+        kafkaStreams2.close();
+        final AtomicInteger usedVersion5 = new AtomicInteger();
+        final KafkaStreams kafkaStreams5 = buildFutureStreams(usedVersion5);
+        startSync(kafkaStreams5);
+        assertThat(usedVersion5.get(), is(LATEST_SUPPORTED_VERSION));
+
+        // third roll, upgrade complete
+        kafkaStreams3.close();
+        final AtomicInteger usedVersion6 = new AtomicInteger();
+        final KafkaStreams kafkaStreams6 = buildFutureStreams(usedVersion6);
+        startSync(kafkaStreams6);
+        assertThat(usedVersion6.get(), is(LATEST_SUPPORTED_VERSION + 1));
```

Comment:
ping @ableegoldman to also take a look at this test logic, are we narrowing down the path sufficiently?